Add new slave lag throttler

jasonhl opened the pull request:
An LHM throttler which tries to query any directly connected DB slaves, verifies their replication lag and adjusts the sleep period in between strides based on the amount of lag present. The sleep period will double each time the lag detected is greater than the desired threshold (up to a maximum).

jasonhl commented:
@camilo @arthurnn @SingerWang: I'm not done with this yet -- it needs more testing and probably some cleanup -- but it is ready for you guys to take a first pass and point out anything that looks wonky. Can you take a look please?

camilo commented:
oh why not make this PR against  soundcloud/lhm ?

camilo commented on lib/lhm/invoker.rb:
```diff
@@ -63,7 +63,7 @@ def normalize_options(options)
 
       if options[:throttler]
-        options[:throttler] = Throttler::Factory.create_throttler(*options[:throttler])
+        options[:throttler] = Throttler::Factory.create_throttler(*options[:throttler].merge({:connection => @connection}))
```
hmrr this will needlessly pass the connection onto say the time throttler, not sure how to make it better. Maybe move away from open option hashes and pass arguments packaged as ostructs with proper names or even classes, might not matter in the end is just a bit unclean. I think is mostly because of the design of the factory.

You can always merge connection only on if the throttler is a lag throttler OR have a class method asking, do you need a connection. Alternatively you could say this does not matter.

jasonhl replied:
Yeah, I messed around with this for a bit -- it's the factory getting in the way. It enforces the fact that all Throttler creation needs to have the same signature, so an open hash makes sense.

I'll look at it a bit more to see if I can come up with something cleaner but I'm not going to kill myself trying to make it elegant.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+        get_slave_hosts.each do |slave|
+          slave_connection = connection_to(slave)
+          result = slave_connection.execute("SELECT MAX(GREATEST(0, UNIX_TIMESTAMP() - UNIX_TIMESTAMP(ts) - 1)) AS lag FROM meta.heartbeat")
+          results.each do |row|
```
can use map and assign

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+        lags = [0]
+        get_slave_hosts.each do |slave|
+          slave_connection = connection_to(slave)
+          result = slave_connection.execute("SELECT MAX(GREATEST(0, UNIX_TIMESTAMP() - UNIX_TIMESTAMP(ts) - 1)) AS lag FROM meta.heartbeat")
```
this query and the one to get the slaves could be constants as well since they are well constant :P

camilo replied:
I think we should rescue any connection exceptions here AND re-raise as our own errors so it is easy to signal what went wrong, "Error: impossible to connect to salves aborting because scary"

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+   
+      def get_slaves
+        list = []
+        return list if @connection.nil? || !@connection.respond_to?(:execute)
```
IMO this check should happen in initialize and raise an ArgumentErr

jasonhl replied:
Yes it should. And that's the way I had it. But... factory and tests. Will try again.

jasonhl replied:
Failed to get this to work. Don't think it's a huge deal -- when the slave list is empty, the answer to "how much slave lag is there?" is zero. 

If you feel strongly about it, I'll try again but leaving this for now.

camilo replied:
hmm wat, how does the connection being nil or not knowing how to execute means that  there is no lag?

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+        list
+      end  
+  
+      def max_slave_lag
```
I'd rename this `max_current_slave_lag` so is not confused with say the ceiling we accept, makes for MOAR easily read `throttle_seconds`

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+          @timeout_seconds
+        end
+
+        @timeout_seconds
```
I think you can skip this return and it'll work the same.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+      def throttle_seconds
+        lag = max_slave_lag
+
+        if lag > @allowed_lag 
```
We should add logging in both cases increasing and decreasing the logging.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+        @timeout_seconds
+      end
+
+      def get_slave_hosts
```
I'd drop the get_ from this name.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,88 @@
+      def get_slave_hosts
+        slaves = []
+        slave_hosts = get_slaves
+        slave_hosts.each do |slave_host|
```
again I think map would work here.

jasonhl commented:
> oh why not make this PR against soundcloud/lhm ?

Because I don't know how to do that without pushing a branch, which I have no permission to do on that repo.

arthurnn commented:
@jasonhl here is the link https://github.com/soundcloud/lhm/compare/master...jasonhl:slave_lag_throttler

jasonhl commented:
@camilo: First round of review comments addressed. Added a bunch more tests and ran locally to confirm that slave lag is being detected and corrected for.

@camilo, @arthurnn: ready for another round of review please.

arthurnn commented on spec/unit/throttler/slave_lag_spec.rb:
```diff
@@ -0,0 +1,102 @@
+
+      it 'does not alter the currently set timeout' do
+        timeout = @throttler.timeout_seconds
+        @throttler.send(:throttle_seconds).must_equal timeout
```
lets use `assert_equal`.

camilo replied:
I like assert better too, but isn't this the rspec way?

arthurnn replied:
we dont use rspec in this project.. this is minitest..

camilo replied:
It says spec all over!! What is this sorcery ?... ( spec adapter I imagine)

> On Mar 13, 2015, at 9:06 AM, Arthur Nogueira Neves <notifications@github.com> wrote:
> 
> In spec/unit/throttler/slave_lag_spec.rb:
> 
> > +
> > +  before :each do
> > +    @throttler = Lhm::Throttler::SlaveLag.new
> > +  end
> > +
> > +  describe '#throttle_seconds' do
> > +    describe 'with no slave lag' do
> > +      before do
> > +        def @throttler.max_current_slave_lag
> > +          0
> > +        end
> > +      end
> > +
> > +      it 'does not alter the currently set timeout' do
> > +        timeout = @throttler.timeout_seconds
> > +        @throttler.send(:throttle_seconds).must_equal timeout
> we dont use rspec in this project.. this is minitest..
> 
> —
> Reply to this email directly or view it on GitHub.
>

arthurnn replied:
Yeah, Minitest has a spec module.

arthurnn commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,90 @@
+      end
+
+      def connection_to(host)
+        Mysql2::Client.new( 
```
can we use a active record connection here, instead of using the driver directly ?

camilo replied:
https://github.com/jasonhl/lhm/blob/slave_lag_throttler/lib/lhm/connection.rb <--- one of these maybs so it all looks the same

arthurnn replied:
:+1: to what camilo said

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,90 @@
+        results = @connection.execute(SQL_SELECT_SLAVE_HOSTS)
+        results.each do |row|
+          list.push(row[0])
+        end
```
oh map here too

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,90 @@
+      end
+
+      def execute
+        sleep throttle_seconds 
```
sleep(throttle_seconds)

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,90 @@
+            result = slave_connection.query(SQL_SELECT_MAX_SLAVE_LAG)
+            result.each do |row|
+              lags.push(row["Seconds_Behind_Master"])
+            end
```
map :P

jasonhl replied:
How would map work here? I'm pushing onto an array which is built up over the outer loop iteration.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,90 @@
+            raise Lhm::Error, "Unable to connect to slave to determine slave lag. Migration aborting."
+          end
+        end
+        lags.push(0).max
```
Hmm pushing 0 is interesting... I guess is equiv to  lags.max.to_i

jasonhl replied:
Not quite... if there's ever some weird negative value returned, pushing 0 ensures it will never return a negative.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,90 @@
+            slave_connection = connection_to(slave)
+            result = slave_connection.query(SQL_SELECT_MAX_SLAVE_LAG)
+            result.each do |row|
+              lags.push(row["Seconds_Behind_Master"])
```
can this return crazy ?

```
irb(main):007:0> ['1', 1].max
ArgumentError: comparison of Fixnum with String failed
```

or even something more exotic ?

jasonhl replied:
I'll to_i it, but I don't think it can return a value other than a number. The docs say it can return a NULL if some of the slave threads are borked, but in that case we're screwed anyway.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,81 @@
+            result = conn.execute(SQL_SELECT_MAX_SLAVE_LAG)
+            result.each(:as => :hash) {|row| lags.push(row["Seconds_Behind_Master"].to_i) }
+          rescue Error
+            raise Lhm::Error, "Unable to connect and/or query slave to determine slave lag. Migration aborting."
```
IMO the inner begin/rescue block could be a private method for SLA.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,81 @@
+                       :password => ActiveRecord::Base.connection_config[:password], 
+                       :database => ActiveRecord::Base.connection_config[:database]
+                     }
+            conn = ActiveRecord::Base.send(adapter_method, config) 
```
Why not use an Lhm::Connection obj?

arthurnn replied:
@camilo I am start to think we can remove the Lhm::Connection abstraction, as we only work with AR now.
So I would not bother about it in here.

camilo replied:
kk, and yes at this point that abstraction is useless

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,81 @@
+      def get_slaves
+        return [] if @connection.nil? || !@connection.respond_to?(:execute)
+        results = @connection.execute(SQL_SELECT_SLAVE_HOSTS)
+        results.map{ |row| row[0] }
```
can be just, and avoid results completely  

@connection.execute(SQL_SELECT_SLAVE_HOSTS).map{&:first}

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,81 @@
+        @timeout_seconds = DEFAULT_TIMEOUT
+        @stride = options[:stride] || DEFAULT_STRIDE
+        @allowed_lag = options[:allowed_lag] || DEFAULT_MAX_ALLOWED_LAG
+        @connection = options[:connection]
```
is this connection an Lhm::Connection or an AR one? ... why not just do the 

```
raise ArgumentError, 'lol connection' unless @connection && !@connection.respond_to?(:execute)
```

I know you explained why but I kinda did not understand.

jasonhl commented:
Squashed commits and fixed up all review comments.

I think this is ready to go now :-)

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,82 @@
+      private
+
+      SQL_SELECT_SLAVE_HOSTS = "SELECT host FROM information_schema.processlist WHERE command='Binlog Dump'"
+      SQL_SELECT_MAX_SLAVE_LAG = "SHOW SLAVE STATUS"
```
This constants are still public even if declared here methinks

jasonhl replied:
Made them private.

jasonhl commented:
Another round complete. Not going to jinx it. Have at it.

jasonhl commented:
And thanks for the reviews and help @camilo and @arthurnn. I learn lots every time.

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,87 @@
+        lags = [0]
+        slave_hosts.each do |slave|
+          lags.concat(slave_lag(slave))
+        end
```
`slave.map { |s|m slave_lag(s)  }.max`

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,87 @@
+      attr_accessor :allowed_lag
+
+      def initialize(options = {})
+        raise ArgumentError, "You must provide a valid :connection option when using the slave lag throttler" unless options[:connection]
```
check the ` respond_to? :execute ` ?

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,87 @@
+                 }
+        conn = Lhm::Connection.new(ActiveRecord::Base.send(adapter_method, config))
+        result = conn.execute(SQL_SELECT_MAX_SLAVE_LAG)
+        result.each(:as => :hash) {|row| lags.push( row["Seconds_Behind_Master"].to_i ) }
```
`lags = result.map(:as => :hash) {|row| lags.push( row["Seconds_Behind_Master"].to_i ) }`

jasonhl replied:
Sure, but `result.map(:as => :hash)` is not a thing.

camilo replied:
`result.each(:as => :hash).map  ` then :P

arthurnn commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -0,0 +1,87 @@
+      def slave_hosts
+        slave_hosts = get_slaves
+        slaves = slave_hosts.map { |slave_host| slave_host.partition(":")[0] }
+        slaves.delete_if { |slave| slave == "localhost" || slave == "127.0.0.1" }
```
How about if the the slave is running on the same server but on different ports? I guess this is ignoring the port .
Also I would say we need to allow localhost , so we can test the integration .

jasonhl replied:
If the slave is running on the same host, we have no way of figuring out the port. The master does not know which port the slave uses to accept requests.

So, the assumption is made that the slave is running on a different host with the same port as master (note the way the connection to the slave is constructed -- we copy the port from master). This is the way Shopify is configured and the only way this throttler will work.

If other cases need to be supported, there would have to be some configuration passed along which would override this "dynamic" way of doing it. But that's for another day :-)

As for the integration tests, I simply overwrote the method to test that the queries work as designed: https://github.com/jasonhl/lhm/pull/1/files#diff-6452a30f664ce4951119752c0f2bff8dR92.

arthurnn replied:
ok this make sense, I can add that extra config.
I wonder how `pt-online-schema-change` handles that? do you know @camilo ?

camilo replied:
I know who knows, @sroysen @singerwang ping

jasonhl commented:
Going off on vacation for the next week. I think this is ready to merge. @camilo -- if it looks all good to you, please grab it and PR against upstream so @arthurnn can merge.

If there's something blocking, I can do it when I get back.
